### PR TITLE
libs-scala: Move `SourceQueueResourceOwner` from the Integration Kit

### PR DESCRIPTION
**Pull request by hubert-da**
Not one-to-one tho, so please check if these changes make sense.
I also checked current usages of `Source.queue` and none of them would be reasonable to migrate.
It is going to be used solely by the Integration Kit, at least for now.

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description
- [ ] If you mean to change the status of a component, please make sure you keep [the Component Status page](https://github.com/digital-asset/daml/blob/main/docs/source/support/component-statuses.rst) up to date.

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala [hunk 1]
```diff
@@ -4,9 +4,8 @@
 package com.daml.resources.akka
 
 import java.util.concurrent.atomic.AtomicBoolean
-
```
Comment (SamirTalwar-DA) [inline libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala:7]: Please configure your editor to leave this newline alone.

File: libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+import scala.concurrent.Future
+
+class SourceQueueResourceOwner[T, Context: HasExecutionContext](
+    queueGraph: RunnableGraph[SourceQueueWithComplete[T]]
+)(implicit
+    materializer: Materializer
+) extends AbstractResourceOwner[Context, SourceQueue[T]] {
+  override def acquire()(implicit context: Context): Resource[Context, SourceQueue[T]] =
+    ReleasableResource(Future(queueGraph.run()))(queue =>
```
Comment (SamirTalwar-DA) [inline libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala:18]: `Resource.apply` delegates to `ReleasableResource`.

```suggestion
    Resource(Future(queueGraph.run()))(queue =>
```

File: libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+      {
+        queue.complete()
+        queue.watchCompletion()
+      }.map(_ => ())
```
Comment (SamirTalwar-DA) [inline libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala:22]:
```suggestion
      queue.complete()
      queue.watchCompletion().map(_ => ())
```

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala [hunk 4]
```diff
@@ -104,4 +103,32 @@ final class AkkaResourceOwnerSpec extends AsyncWordSpec with Matchers {
...
+              .queue[Int](10, OverflowStrategy.backpressure)
+              .to(Sink.foreach(number = _))
+          )(materializer)
+          .acquire()
+      } yield sourceQueue
+
+      for {
+        queue <- resource.asFuture
+        offerResult <- queue.offer(123)
+        _ <- resource.release()
```
Comment (SamirTalwar-DA) [inline libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala:127]: This seems to be the pattern in this test file, but this is prone to leaks if tests fail. Do you think you can change the file to use `ResourceOwner#use` (in a separate PR)?

**Review by SamirTalwar-DA — APPROVED** [review]

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala (hunk 4 above)
Comment (hubert-da) [reply libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala:127]: Yes, I was thinking about this, will try.

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala (hunk 1 above)
Comment (hubert-da) [reply libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala:7]: Weird, it was working fine previously.

File: libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala (hunk 2 above)
Comment (hubert-da) [reply libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala:18]: I think the resource you're referring to comes from `ledger/ledger-resources`.  The owners from this package use the `ReleasableResource`, are you fine to keep it, @SamirTalwar-DA?

Comment (SamirTalwar-DA) [reply libs-scala/resources-akka/src/main/scala/com/digitalasset/resources/akka/SourceQueueResourceOwner.scala]: Yeah, sure, if it's easier.

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala (hunk 4 above)
Comment (hubert-da) [reply libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala]: Sorry @SamirTalwar-DA, as there was some flakiness, I decided to put it into this PR. Please take a look.

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala [hunk 5]
```diff
@@ -35,56 +33,60 @@ final class AkkaResourceOwnerSpec extends AsyncWordSpec with Matchers {
...
+      val actorSystemReference = new AtomicReference[ActorSystem]()
+
+      val resultFuture = resourceOwner
+        .use { case (actorSystem, actor) =>
+          actorSystemReference.set(actorSystem)
+          actor ! 7
+          testPromise.future
+        }
+
+      blocking {
```
Comment (SamirTalwar-DA) [inline libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala:50]: If you switch `AnyWordSpec` to `AsyncWordSpec`, you will be able to do this (here and below), as you can return a `Future[Assertion]`.

```scala
    resourceOwner
      .use { (actorSystem, actor) =>
        actor ! 7
        testPromise.future.map { result =>
          result should be(7)
          actorSystem
        } 
      }
      .map { actorSystem =>
        an[IllegalStateException] should be thrownBy actorSystemReference.get.actorOf(
          Props(new TestActor)
        )
      }
```

I find this a lot simpler, as there's no calls to `Await.result` or `AtomicReference#set`.

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala (hunk 4 above)
Comment (hubert-da) [reply libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala]: I reported the `lsp-tests` flakiness on Slack, btw.

File: libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala (hunk 5 above)
Comment (hubert-da) [reply libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala:50]: Ah, you're right, `release` is being called earlier, gimme a sec.

Comment (hubert-da) [reply libs-scala/resources-akka/src/test/suite/scala/com/digitalasset/resources/akka/AkkaResourceOwnerSpec.scala]: Done

**Review by SamirTalwar-DA — APPROVED** [review]

**Pull request merged**